hw/ip/otbn/util/shared/information_flow_analysis.py

```diff
@@ -0,0 +1,598 @@
+    out = (used_constants, return_iflow, program_end_iflow, common_constants,
+           control_deps)
+    _get_iflow_cache_update(start_pc, start_constants, out, cache)
+    return out
+
+
+def check_acyclic(graph: ControlGraph) -> None:
+    '''Checks for (non LOOP/LOOPI) cycles in control-flow graph.
+
+    If there are such cycles, we need to raise an error and not proceed; the
+    control-flow traversal would infinite-loop.
+    '''
+    cycles = graph.get_cycles(graph.start)
+    if len(cycles) != 0:
+        msg = [
+            'One or more cycles found in control-flow graph at the following PCs:'
+        ]
+        for pc, links in cycles.items():
+            msg.append('{:#x} <-> {}'.format(
+                pc, ','.join(['{:#x}'.format(l) for l in links])))
+        msg.append('Analyzing cyclic control flow outside of LOOP/LOOPI '
+                   'instructions is not currently supported.')
+        raise ValueError('\n'.join(msg))
+    return
+
+
+def get_subroutine_iflow(
+    program: OTBNProgram, graph: ControlGraph, subroutine_name: str
+) -> Tuple[Optional[InformationFlowGraph], Optional[InformationFlowGraph],
+           Dict[str, Set[int]]]:
+    '''Gets the information-flow graphs for the subroutine.
+
+    Returns threee items:
+    1. The combined information flow for control paths that return from the
+       subroutine to its caller (None if there are no such paths)
+    2. The combined information flow for control paths ending in the end of the
+       program (e.g. ECALL) (None if there are no such paths)
+    3. The information-flow nodes whose values at the start of the subroutine
+       influence its control flow.
+    '''
+    check_acyclic(graph)
+    start_pc = program.get_pc_at_symbol(subroutine_name)
+    _, ret_iflow, end_iflow, _, control_deps = _get_iflow(
+        program, graph, start_pc, {'x0': 0}, None, {})
+    if ret_iflow is None and end_iflow is None:
+        raise ValueError('Could not find any complete control-flow paths when '
+                         'analyzing subroutine.')
+    return ret_iflow, end_iflow, control_deps
+
+
+def get_program_iflow(
+        program: OTBNProgram, graph: ControlGraph
+) -> Tuple[InformationFlowGraph, Dict[str, Set[int]]]:
+    '''Gets the information-flow graph for the whole program.
+
+    Also returns the information-flow nodes whose values at the start of the
+    program influence its control flow.
+    '''
+    check_acyclic(graph)
+    _, ret_iflow, end_iflow, _, control_deps = _get_iflow(
+        program, graph, program.min_pc(), {'x0': 0}, None, {})
+    if ret_iflow is not None:
+        # No paths from imem_start should end in RET
+        raise ValueError('Unexpected information flow for paths ending in RET '
+                         'when analyzing whole program.')
+    if end_iflow is None:
+        raise ValueError('Could not find any complete control-flow paths when '
+                         'analyzing whole program.')
```

Comment:
Is this "Your program will hang" ? If so, maybe worth saying so explicitly?

Reply:
This is sort of a very weird an unexpected situation as the code is currently written. Since cyclic control-flow graphs are rejected and loops are handled, that means you somehow have a program that doesn't have cycles and also never reaches the end of IMEM or an `ecall` -- which really shouldn't be possible, so it probably means there's a bug in the analysis. I'll change this to an assertion to better reflect that this is really unexpected.

Reply:
Ah, cool. Thanks for explaining.